src/handshake/client.rs

```diff
@@ -108,6 +108,10 @@ fn generate_request(request: Request, key: &str) -> Result<Vec<u8>> {
         return Err(Error::Url(UrlError::EmptyHostName));
     }
 
+    if let Some(v) = request.headers().get(header::HOST) {
+        host = unsafe { std::str::from_utf8_unchecked(v.as_ref()) };
+    }
+
```

Comment:
But what's the point of parsing the `host` value if we're going to overwrite it in presence of the custom user header? What I mean is that the same logic could then be rewritten like this:

```rust
if let Some(user_host_header) = requests.headers().get(header::HOST) {
    host = ...
} else if let Some(idx) = authority.find('@') {
   ....
}
```

So that we don't need to have a `mut host` and the code is a bit more safe.

In addition, I don't think introducing an `unsafe` block here would be a good decision as the same logic could be written safely.

Speaking more broadly I'm not sure that this change would make sense in general because of the RFC 6455 that states that:

```
       <...> the URI to
       pass the function MUST be constructed from /host/, /port/,
       /resource name/, and the /secure/ flag using the definition of a
       WebSocket URI as given in Section 3.

<...>

   4.   The request MUST contain a |Host| header field whose value
        contains /host/ plus optionally ":" followed by /port/ (when not
        using the default port).
```

I.e. it does not seem like we're allowed to overwrite the `Host` here.

There is [section 12](https://datatracker.ietf.org/doc/html/rfc6455#section-12) in the RFC but I'm not sure if it's the one that you had in mind to apply here:

```
12.  Using the WebSocket Protocol from Other Specifications

   The /host/, /port/, /resource name/, and /secure/ flag are usually
   obtained from a URI using the steps to parse a WebSocket URI's
   components.  These steps fail if the URI does not specify a
   WebSocket.
```

CC @jxs

Reply:
my scene is that we want to  dial to a shared endpoint, then the shared endpoint can dispatch the differently upstream by `host`.

Reply:
But it does seem to contradict the RFC, doesn't it?

Reply:
but some http clients, for example hyper allow users to overwrite `host` in headers . maybe it's a de facto standard

https://github.com/hyperium/hyper/blob/6393a0cc031f504dbd50eb4b021bae51e21f6140/src/client/client.rs#L248-L259
```rs
            if self.config.set_host {
                let uri = req.uri().clone();
                req.headers_mut().entry(HOST).or_insert_with(|| {
                    let hostname = uri.host().expect("authority implies host");
                    if let Some(port) = get_non_default_port(&uri) {
                        let s = format!("{}:{}", hostname, port);
                        HeaderValue::from_str(&s)
                    } else {
                        HeaderValue::from_str(hostname)
                    }
                    .expect("uri host is valid header value")
                });

```

Reply:
Hm, that's interesting. I wonder if any other WebSocket libraries allow it. We could theoretically allow it, it's not a big deal to handle it in the code. What bothered me a little bit is that it's effectively against the RFC sections that I mentioned in my comments. Do I miss something?

Reply:
gorilla websocket client allow it. https://github.com/gorilla/websocket/blob/e8629af678b7fe13f35dff5e197de93b4148a909/client.go#L207-L224.

hm, it looks like RFC doesn't show `Host` can be allowed to be different with original URI. I think it's more likely to be a de facto standard. 
there some discuss
https://github.com/curl/curl/issues/3664